src/NServiceBus.Core/Routing/DistributionStrategy.cs

```diff
@@ -24,6 +27,14 @@ protected DistributionStrategy(string endpoint, DistributionStrategyScope scope)
         public abstract string SelectReceiver(string[] receiverAddresses);
 
         /// <summary>
+        /// Selects a destination instance for a message from all known addresses of a logical endpoint.
+        /// </summary>
+        public virtual Task<string> SelectReceiver(string[] receiverAddresses, IOutgoingContext outgoingContext)
```

Comment:
I am not sure I like this approach. We end up with an abstract class that has two overridable methods and it up to the doco/comments to ensure the user overrides only one of them. What should he expect if he overrides both? -- can't tell just looking at the API. I would be confused if my `sync` version is called when I also override the optional `async` one. Or should I call it myself?

Would it be a breaking change if we had `AsyncDistributionStrategy` as a new base class for `DistributionStrategy` ? then the new `async` method can be defined in the base class and the user inherits either directly from the `async` one if he needs to do IO or he inherits from a sync one if he does not. The sync would override with `sealed` to ensure only the sync method is overridable at this point.